### DISCUSSION TITLE
python duckdb.connect is really slow

### DISCUSSION THREAD

**Discussion by slederman** [duckdb/duckdb#4403] [Q&A]
I've got a duckdb database of 227 million rows that takes up nearly 700GB. I'm using the python duckdb package and con = duckdb.connect(database='mydb.duckdb') to make a connection. That connecting takes 15 seconds. Is there any way to make it go a lot faster?

**Comment by pdet** [comment]
Are you defining primary/foreign keys?

**Reply by slederman** [reply]
Good question. I'm having duckdb load parquet files into a table and I haven't defined a schema. I'll look at specifying a schema with a primary key.

**Reply by pdet** [reply]
The reason I ask is because index storage was recently implemented (not yet released, only available in the master branch), and was a common cause for slow database loading times, since the index would be rebuilt every time. However, if you are not defining them, that should not be the cause.

Can you share the script you are using to create the database and reload it?

**Comment by slederman** [comment]
The script has lots of processing that's not relevant to this discussion but here's the gist of it.

1. Create a python list of dicts, each dict is one row of data. 
2. Create a pyarrow table from the list: table = pa.Table.from_pylist(the_list). Let pyarrow infer the schema.
3. Write the pyarrow table to parquet: pq.write_table(table,parquet_file)
4. con = duckdb.connect(database='my_duckdb_file')
5. insert into duckdb_table select * from parquet_file. Repeat for each parquet file, closing and reopening the connection for each file.

**Comment by Mytherin** [comment]
This is likely caused by us not lazily loading the metadata. With a 700GB database even just constructing all the row groups + the column data objects and loading statistics will be expensive. We need to move towards doing this lazily and potentially in parallel.

**Reply by slederman** [reply]
Thanks for the information. Should I turn this discussion into an issue?
